torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -223,19 +223,9 @@ class ProcessGroupNCCL : public ProcessGroup {
           "FutureNCCL only supports single-process single-device mode.");
     }
 
-    // This constructor is used by then callback, it skips setting the value at
-    // the beginning. Later, the value will be set using markCompleted with the
-    // return value of callback.
-    explicit FutureNCCL(
-        c10::DeviceIndex deviceIndex,
-        std::shared_ptr<std::vector<at::cuda::CUDAEvent>> cudaEvents)
+    explicit FutureNCCL(c10::DeviceIndex deviceIndex)
         : at::ivalue::Future(c10::ListType::create(c10::TensorType::get())),
-          deviceIndex_(deviceIndex),
-          cudaEvents_(std::move(cudaEvents)) {
```

Comment:
this means `C++` code can no longer call `completed()` after constructing `FutureNCCL` using this ctor?

Reply:
I see what you mean and, yes, you are right. Note however that the FutureNCCL class (even before I started reworking it) relied very strongly on the implicit assumption that instances would be marked complete immediately upon construction. Without that assumption, the behavior of FutureNCCL would be fundamentally wrong (e.g., it always calls the callbacks immediately inline). So far, I am still relying on that assumption (I'll relax it in #48505). Hence, this second constructor is only called by the `then()` method, which however then runs the callback inline and thus when it returns the future to the user it's already complete. Thus it's still correct (even for C++ users) to always return true from `completed()`.

If you'd like some additional safety here, one thing I can do is to make this second constructor private, so that we can really be sure that this will only be used in the `then()` method. However, we'd have to make it public again in #48505.